src/amdgpu_libdrm.cpp

```diff
@@ -0,0 +1,112 @@
+#include <deque>
+#include <mutex>
+#include <thread>
+#include <chrono>
+#include <xf86drm.h>
+#include <libdrm/amdgpu.h>
+#include <spdlog/spdlog.h>
+#include <fcntl.h>
+#include "gpu.h"
+
+#include "amdgpu_libdrm.h"
+
+std::string dri_device_path;
+bool do_libdrm_sampling = false;
+
+std::deque<struct libdrm_sample> sample_buf(LIBDRM_SAMPLE_BUF_SIZE, {0});
+std::mutex sample_buf_m;
+
+amdgpu_device_handle amdgpu_handle;
+
+static void libdrm_do_sample(libdrm_sample *sample) {
+    uint32_t registers;
+    amdgpu_read_mm_registers(amdgpu_handle, LIBDRM_GRBM_STATUS / 4, 1, 0xffffffff, 0, &registers);
+
+    if (registers & LIBDRM_GRBM_BUSY_BIT) sample->busy_bit = true;
+}
+
+static void libdrm_thread() {
+    while (true) {
+        auto start_time = std::chrono::system_clock::now().time_since_epoch();
+
+        struct libdrm_sample sample {0};
+        libdrm_do_sample(&sample);
+        
+        sample_buf_m.lock();
+        sample_buf.pop_front();
+        sample_buf.push_back(sample);
+        sample_buf_m.unlock();
+
+        auto end_time = std::chrono::system_clock::now().time_since_epoch();
+        auto sleep_duration = std::chrono::microseconds(LIBDRM_SAMPLE_DELAY) - (end_time - start_time);
+        if (sleep_duration > std::chrono::nanoseconds(0)) {
+            std::this_thread::sleep_for(sleep_duration);
+        }
+    }
+}
+
+static int libdrm_initialize() {
+    drmDevicePtr devices[LIBDRM_MAX_DEVICES];
+    int device_count = drmGetDevices2(0, devices, LIBDRM_MAX_DEVICES);
+    if (device_count < 0) {
+        SPDLOG_ERROR("drmGetDevices2 failed");       
+        return -1;
+    }
+
+    char *renderd_node = nullptr;
+    for (int i = 0; i < device_count; i++) {
+        if ((devices[i]->available_nodes & 0b101) != 0b101) {
```

Comment:
Using `DRM_NODE_PRIMARY | DRM_NODE_RENDER` over the binary numbers, makes this easier to read.

Reply:
`DRM_NODE_PRIMARY | DRM_NODE_RENDER` results in `0b10` and not `0b101`. I couldn't find any way I could use that, maybe because I'm really bad at bitmasks

Reply:
Note to self - don't review before coffee is drunk:

```c
#define THE_MASK (1 << DRM_NODE_PRIMARY | 1 << DRM_NODE_RENDER)

if ((devices[i]->available_nodes & THE_MASK != THE_MASK ) {
```

Reply:
Thanks